Cluster Health Check

mattmattox opened the pull request:
This script is designed to capture the following items from a Kubernetes cluster. This is based on the Log-Collector script.

- Cluster
  - kubectl cluster-info
  - kubectl get nodes
- Nodes
  - kubectl get nodes
  - kubectl top nodes
  - kubectl describe nodes
  - A break down of the number of nodes in each role
  - How many pods are on each node
- Pod Info
  - kubectl get pods -A
  - A break down of pods in each status, (Running, Pending, Failed, Unknown, Completed, CrashLoopBackOff, NodeAffinity, and ImagePullBackOff)
  - A summary of the pods count for each status
  - kubectl top pod
- Storage
  - kubectl get storageclass
  - kubectl get pv
- Namespaced info for Rancher/System namespaces
  - pods
  - services
  - endpoints
  - daemonset
  - deployment
  - replicaset
  - jobs
  - cronjob
  - configmap
  - secret (No -o yaml, just a list)
  - pvc

We then run some tests
- Overlay
  - All nodes ping all nodes
- DNS
  - We resolve kubernetes.default and connect from each node
- kube-apiserver
  - It tests connecting to each kube-apiserver from each node and test via nginx-proxy as well

We then archive everything into a tar.gz file.

dnoland1 commented:
tried this out on a k3s cluster. Noticed the DNS test failed with some errors trying to find tokens:
```
root@ip-10-10-101-182:~# ./cluster-health.sh
2021-02-19 07:41:49: Created /tmp/tmp.LeM38u4TiR
2021-02-19 07:41:49: Verifying cluster access
2021-02-19 07:41:50: Cluster access has been verified
2021-02-19 07:41:50: Collecting information about the cluster...
2021-02-19 07:41:58: Collecting information about the nodes...
2021-02-19 07:42:00: Gathering node counts...
2021-02-19 07:42:04: Node Summary Report
2021-02-19 07:42:04: All nodes: 2
2021-02-19 07:42:04: etcd: 0
2021-02-19 07:42:04: controlplane: 0
2021-02-19 07:42:04: worker: 0
2021-02-19 07:42:04: master: 2
2021-02-19 07:42:04: Pods Per Node
2021-02-19 07:42:05: ip-10-10-101-182 16
2021-02-19 07:42:05: ip-10-10-100-93 13
2021-02-19 07:42:05: Collecting information about the pods...
2021-02-19 07:42:12: Pod Summary Report
2021-02-19 07:42:12: Pods: 27
2021-02-19 07:42:12: Running: 27
2021-02-19 07:42:12: Pending: 0
2021-02-19 07:42:12: Failed: 0
2021-02-19 07:42:12: Unknown: 0
2021-02-19 07:42:12: Completed: 0
2021-02-19 07:42:12: CrashLoopBackOff: 0
2021-02-19 07:42:12: NodeAffinity: 0
2021-02-19 07:42:12: ImagePullBackOff: 0
2021-02-19 07:42:12: NodeAffinity: 0
2021-02-19 07:42:13: Collecting information about storage...
2021-02-19 07:42:15: Collecting information about cattle-dashboards
2021-02-19 07:42:22: Collecting information about cattle-logging-system
2021-02-19 07:42:30: Collecting information about cattle-monitoring-system
2021-02-19 07:42:38: Collecting information about cattle-system
2021-02-19 07:42:46: Collecting information about cis-operator-system
2021-02-19 07:42:54: Collecting information about fleet-system
2021-02-19 07:43:03: Collecting information about ingress-nginx
2021-02-19 07:43:11: Collecting information about kube-node-lease
2021-02-19 07:43:18: Collecting information about kube-public
2021-02-19 07:43:26: Collecting information about kube-system
2021-02-19 07:43:34: Collecting information about local-path-storage
2021-02-19 07:43:42: Collecting information about longhorn-system
2021-02-19 07:43:50: Deploying overlay test containers
daemonset.apps/overlaytest created
2021-02-19 07:43:51: Waiting for rollout to complete...
Waiting for daemon set "overlaytest" rollout to finish: 0 of 2 updated pods are available...
Waiting for daemon set "overlaytest" rollout to finish: 1 of 2 updated pods are available...
daemon set "overlaytest" successfully rolled out
2021-02-19 07:44:29: Starting network overlay test
2021-02-19 07:44:32: Success: overlaytest-7gxb5 on ip-10-10-100-93 can reach pod IP 172.16.2.14 on ip-10-10-100-93
2021-02-19 07:44:33: Success: overlaytest-7gxb5 on ip-10-10-100-93 can reach pod IP 172.16.3.20 on ip-10-10-101-182
2021-02-19 07:44:35: Success: overlaytest-227b4 on ip-10-10-101-182 can reach pod IP 172.16.2.14 on ip-10-10-100-93
2021-02-19 07:44:37: Success: overlaytest-227b4 on ip-10-10-101-182 can reach pod IP 172.16.3.20 on ip-10-10-101-182
2021-02-19 07:44:37: Finished network overlay test
2021-02-19 07:44:37: Cleaning up network overlay test
daemonset.apps "overlaytest" deleted
2021-02-19 07:44:37: Deploying DNS test containers
daemonset.apps/dnstest created
2021-02-19 07:44:38: Waiting for rollout to complete...
Waiting for daemon set "dnstest" rollout to finish: 0 of 2 updated pods are available...
Waiting for daemon set "dnstest" rollout to finish: 1 of 2 updated pods are available...
daemon set "dnstest" successfully rolled out
2021-02-19 07:44:40: Starting cluster DNS test
cat: /var/run/secrets/kubernetes.io/serviceaccount/token: No such file or directory
command terminated with exit code 77
2021-02-19 07:44:41: Failure: DNS is not working correctly for dnstest-5r8w4 on ip-10-10-100-93
cat: /var/run/secrets/kubernetes.io/serviceaccount/token: No such file or directory
command terminated with exit code 77
2021-02-19 07:44:42: Failure: DNS is not working correctly for dnstest-2lnbn on ip-10-10-101-182
2021-02-19 07:44:42: Finished DNS dns test
2021-02-19 07:44:42: Cleaning up DNS dns test
daemonset.apps "dnstest" deleted
2021-02-19 07:44:43: Created /tmp/ip-10-10-101-182-2021-02-19_07_44_42.tar.gz
2021-02-19 07:44:43: Removing /tmp/tmp.LeM38u4TiR
```

I also noticed the image used for DNS test is over 400 MB:
```
docker.io/rancher/hyperkube                           v1.19.7-rancher1    3f163568acb72       418MB
```
Is there a smaller image that could be used?

mattmattox commented:
@dnoland1 The DNS test issue is because the default service account isn't created in k3s clusters. (Testing this) I think I can work around this by deploying a sa and role.

For the hyperkube image issue, the reason I'm using hyperkube is that we already deploy it on most nodes as part of RKE. So I was just reusing that image. Also, it allows me to detect if this is air-gapped because I'm grabbing the full image path.

Does it make more sense to switch to a custom alpine-based image and add some air-gapped flags to the script?

dnoland1 commented:
@mattmattox Guess my concern is if someone runs this on a large k3s cluster, the dnstest daemonset requires a 400MB+ download to each node. It might be slow to deploy and hammer the network (40GB for a 100 node cluster). Think it's worth considering leveraging a small image or at least consider this for a version 2 of the health check script.

mattmattox commented:
@dnoland1 I'm going to switch over to using the swiss-army-knife image. But that is pending  https://github.com/rancherlabs/swiss-army-knife/pull/2

I'm currently working on switching swiss-army-knife from a ubuntu-based to an alpine-based image.

leodotcloud commented:
@mattmattox probably we should use a separate image for edge/iot devices. Let's keep ubuntu as the base for the current one.

mattmattox commented:
@dnoland1 @leodotcloud OK, I'll leave the swiss-army-knife as-is (Maybe we'll look at making a lite version in the future)

I switched the script over to using rancherlabs/swiss-army-knife and added an override flag to allow the user to set their own image. This mainly for air-gapped environments.

dnoland1 reviewed (changes requested):
Tested latest on k3s cluster and received errors. See below:
```
root@ip-10-10-101-133:/tmp# ./cluster-health.sh
2021-03-08 21:31:02: Created /tmp/tmp.ISth2TnL33
2021-03-08 21:31:02: Verifying cluster access
2021-03-08 21:31:03: Cluster access has been verified
2021-03-08 21:31:03: Collecting information about the cluster...
2021-03-08 21:31:32: Collecting information about the nodes...
2021-03-08 21:31:34: Gathering node counts...
2021-03-08 21:31:37: Node Summary Report
2021-03-08 21:31:37: All nodes: 2
2021-03-08 21:31:37: etcd: 0
2021-03-08 21:31:37: controlplane: 0
2021-03-08 21:31:37: worker: 0
2021-03-08 21:31:37: master: 2
2021-03-08 21:31:37: Pods Per Node
2021-03-08 21:31:38: ip-10-10-101-133 15
2021-03-08 21:31:38: ip-10-10-100-31 19
2021-03-08 21:31:38: Collecting information about the pods...
2021-03-08 21:31:46: Pod Summary Report
2021-03-08 21:31:46: Pods: 32
2021-03-08 21:31:46: Running: 28
2021-03-08 21:31:46: Pending: 0
2021-03-08 21:31:46: Failed: 0
2021-03-08 21:31:46: Unknown: 0
2021-03-08 21:31:46: Completed: 4
2021-03-08 21:31:46: CrashLoopBackOff: 0
2021-03-08 21:31:46: NodeAffinity: 0
2021-03-08 21:31:46: ImagePullBackOff: 0
2021-03-08 21:31:46: NodeAffinity: 0
2021-03-08 21:31:47: Collecting information about storage...
2021-03-08 21:31:48: Collecting information about cattle-dashboards
2021-03-08 21:31:56: Collecting information about cattle-logging-system
2021-03-08 21:32:03: Collecting information about cattle-monitoring-system
2021-03-08 21:32:11: Collecting information about cattle-system
2021-03-08 21:32:19: Collecting information about cis-operator-system
2021-03-08 21:32:27: Collecting information about fleet-system
2021-03-08 21:32:34: Collecting information about ingress-nginx
2021-03-08 21:32:43: Collecting information about kube-node-lease
2021-03-08 21:32:51: Collecting information about kube-public
2021-03-08 21:33:00: Collecting information about kube-system
2021-03-08 21:33:07: Collecting information about local-path-storage
2021-03-08 21:33:15: Collecting information about longhorn-system
2021-03-08 21:33:22: Deploying serviceaccount
serviceaccount/cluster-health-check created
2021-03-08 21:33:23: Deploying clusterrole
clusterrole.rbac.authorization.k8s.io/cluster-health-check created
2021-03-08 21:33:24: Deploying clusterrolebinding
clusterrolebinding.rbac.authorization.k8s.io/cluster-health-check created
2021-03-08 21:33:24: Deploying overlay test containers
The DaemonSet "overlaytest" is invalid: spec.template.spec.containers[0].image: Required value
2021-03-08 21:33:25: Waiting for rollout to complete...
Error from server (NotFound): daemonsets.apps "overlaytest" not found
2021-03-08 21:33:25: Starting network overlay test
2021-03-08 21:33:25: Finished network overlay test
2021-03-08 21:33:25: Cleaning up network overlay test
Error from server (NotFound): daemonsets.apps "overlaytest" not found
2021-03-08 21:33:26: Deploying DNS test containers
The DaemonSet "dnstest" is invalid: spec.template.spec.containers[0].image: Required value
2021-03-08 21:33:26: Waiting for rollout to complete...
Error from server (NotFound): daemonsets.apps "dnstest" not found
2021-03-08 21:33:27: Starting cluster DNS test
2021-03-08 21:33:27: Finished DNS dns test
2021-03-08 21:33:27: Cleaning up DNS dns test
Error from server (NotFound): daemonsets.apps "dnstest" not found
2021-03-08 21:33:27: Deleting clusterrolebinding
clusterrolebinding.rbac.authorization.k8s.io "cluster-health-check" deleted
2021-03-08 21:33:28: Deleting clusterrole
clusterrole.rbac.authorization.k8s.io "cluster-health-check" deleted
2021-03-08 21:33:28: Deleting serviceaccount
serviceaccount "cluster-health-check" deleted
2021-03-08 21:33:33: Created /tmp/ip-10-10-101-133-2021-03-08_21_33_28.tar.gz
2021-03-08 21:33:33: Removing /tmp/tmp.ISth2TnL33
```

mattmattox commented:
@dnoland1 It should be working on k3s cluster only with the only limitation being swiss-army-knife doesn't have an arm64 image (working on this) so I have some selector rules to only deploy it to amd64 nodes.

mattmattox commented:
I also did some rework to only deploy the swiss-army-knife containers once then use the same containers for each test.

leodotcloud commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,519 @@
+      DISK_FULL=1
+  fi
+}
+verify-access() {
```
If the cluster is in a bad state, the `kubectl` command would probably won't go through. It would be good to through an error and bail out than going through the rest of the script.

leodotcloud commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,519 @@
+}
+verify-access() {
+  techo "Verifying cluster access"
+  if [[ ! -z $OVERRIDE_KUBECONFIG ]];
```
Instead of using this environment variable, can you add `--kubeconfig` argument directly to the script? It would provide a better UX.

leodotcloud commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,519 @@
+  done
+  techo "Gathering node counts..."
+  decho "Total number of nodes..."
+  NumberOfNodes=`${KUBECTL_CMD} get nodes --no-headers | wc -l`
```
We should run any `kubectl` command only once in the script. Rerunnning the same command over and over increases the overall runtime of the script. Ideally, if you can grab the info using `-o json` and redirect it to a file, you can then use `jq` commands to gather the necessary information. Also, we should split the logic into two different phases. First one does just collection of the information in json format, the second format would summarize the necessary information.. This would help reduce the overall bundle size too.

mattmattox replied:
I consolidated as much of the kubectl commands as I could. I did run into issues with the role selector for the node section as it looks like jq doesn't like the '/' in the role name.

leodotcloud commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,519 @@
+deploy-swiss-army-knife() {
+  techo "Deploying swiss-army-knife test containers"
+  get-debug-tool-image
+  decho "DebugToolImage: $DebugToolImage"
```
A good programming practice is to stick to a single naming convention. Either ALLCAPS or CamelCase or snakeCase. In general, all variables in shell scripts should be ALLCAPS.

leodotcloud commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,519 @@
+testsonly=0
+cleanup=0
+
+while getopts ":d:s:r:i:tcfhD" opt; do
```
Since you already have various steps split out in different functions, is it possible to call a certain function?

leodotcloud commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,519 @@
+  pods
+  storage
+  Namespaces='cattle-dashboards cattle-logging-system cattle-monitoring-system cattle-system cis-operator-system fleet-system ingress-nginx kube-node-lease kube-public kube-system local-path-storage longhorn-system'
+  for Namespace in $Namespaces
```
Fix naming convention

mattmattox replied:
Fixed https://github.com/rancherlabs/support-tools/pull/99/commits/9d15556bd2131201bee3b1ff6f4fdfa992e9c34c

leodotcloud commented on cluster-health-check/README.md:
```diff
@@ -0,0 +1,29 @@
+The script will create a .tar.gz log collection in /tmp by default, all flags are optional.
+
+## Flags
+
```
Could you add more verbose debug print statements?

leodotcloud commented on cluster-health-check/README.md:
```diff
@@ -0,0 +1,29 @@
+
+All flags are optional
+-d    Output directory for temporary storage and .tar.gz archive (ex: -d /var/tmp)
+-k    Override the kubeconfig (ex: ~/.kube/custom)
```
`--kubeconfig`???

leodotcloud commented on cluster-health-check/README.md:
```diff
@@ -0,0 +1,29 @@
+Usage: cluster-health.sh [ -d <directory> -k ~/.kube/config -i rancherlabs/swiss-army-knife -f -D ]
+
+All flags are optional
+-d    Output directory for temporary storage and .tar.gz archive (ex: -d /var/tmp)
```
`-d` is generally used for debug, could you use `--output-dir` instead?

leodotcloud commented on cluster-health-check/README.md:
```diff
@@ -0,0 +1,29 @@
+
+The script will create a .tar.gz log collection in /tmp by default, all flags are optional.
+
+## Flags
```
IMHO using long form flags is much easier than remembering these short forms. Anyways, it's up to you.

leodotcloud commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,519 @@
+  ${KUBECTL_CMD} cluster-info > $TMPDIR/clusterinfo/cluster-info 2>&1
+  ${KUBECTL_CMD} cluster-info dump > $TMPDIR/clusterinfo/cluster-info-dump 2>&1
+  decho "get nodes"
+  ${KUBECTL_CMD} get nodes -o wide > $TMPDIR/clusterinfo/get-node.wide 2>&1
+  ${KUBECTL_CMD} get nodes -o yaml > $TMPDIR/clusterinfo/get-node-yaml 2>&1
```
Same info being collected twice as mentioned in another comment.

dnoland1 reviewed (commented):
I ran this on a k3s node and it crashed the VM. This is how far it got:
```
root@ip-10-10-101-64:~# ./cluster-health.sh
2021-03-25 00:58:53: Created /tmp/tmp.HxcjGihm2I
2021-03-25 00:58:53: Verifying cluster access
2021-03-25 00:58:54: Cluster access has been verified
2021-03-25 00:58:54: Collecting information about the cluster...
2021-03-25 01:02:59: Detecting the cluster provider...
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
2021-03-25 01:03:00: Collecting information about the nodes...
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
2021-03-25 01:03:05: Gathering node counts...
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
2021-03-25 01:03:10: Node Summary Report
2021-03-25 01:03:10: Total number of nodes: 0
2021-03-25 01:03:10: etcd: 0
2021-03-25 01:03:10: controlplane: 0
2021-03-25 01:03:10: worker: 0
2021-03-25 01:03:10: master: 0
2021-03-25 01:03:10: Pods Per Node
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
2021-03-25 01:03:11: Collecting information about the pods...
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
2021-03-25 01:03:23: Pod Summary Report
2021-03-25 01:03:23: Pods: 0
2021-03-25 01:03:23: Running: 0
2021-03-25 01:03:23: Pending: 0
2021-03-25 01:03:23: Failed: 0
2021-03-25 01:03:23: Unknown: 0
2021-03-25 01:03:23: Completed: 0
2021-03-25 01:03:23: CrashLoopBackOff: 0
2021-03-25 01:03:23: NodeAffinity: 0
2021-03-25 01:03:23: ImagePullBackOff: 0
2021-03-25 01:03:23: NodeAffinity: 0
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
2021-03-25 01:03:24: Collecting information about storage...
2021-03-25 01:03:27: Collecting information about cattle-dashboards
2021-03-25 01:03:38: Collecting information about cattle-logging-system
2021-03-25 01:03:49: Collecting information about cattle-monitoring-system
```

mattmattox reviewed (commented):
@leodotcloud I consolidated as much of the kubectl commands as I could. I did run into issues with the role selector for the node section as it looks like jq doesn't like the '/' in the role name.

dnoland1 commented:
I ran on a vanilla one node RKE2 cluster. The node and pod summary counts seem to be incorrect. Here's the full output:
```
root@ip-10-0-7-10:~# ./cluster-health.sh
2021-05-09 16:07:23: Created /tmp/tmp.LUZxYxiCy9
2021-05-09 16:07:23: Detecting available commands... renice ionoice
2021-05-09 16:07:23: Verifying cluster access
2021-05-09 16:07:23: Cluster access has been verified
2021-05-09 16:07:23: Collecting information about the cluster...
2021-05-09 16:07:24: Detecting the cluster provider...
2021-05-09 16:07:24: Collecting information about the nodes...
W0509 16:07:25.221833   13189 top_node.go:119] Using json format to get metrics. Next release will switch to protocol-buffers, switch early by passing --use-protocol-buffers flag
2021-05-09 16:07:25: Gathering node counts...
2021-05-09 16:07:25: Node Summary Report
2021-05-09 16:07:25: Total number of nodes: 1
2021-05-09 16:07:26: etcd: 4
2021-05-09 16:07:26: controlplane: 3
2021-05-09 16:07:26: worker: 3
2021-05-09 16:07:26: master: 4
2021-05-09 16:07:26: Pods Per Node
2021-05-09 16:07:26: ip-10-0-7-10 18
2021-05-09 16:07:26: Collecting information about the pods...
2021-05-09 16:07:27: Pod Summary Report
2021-05-09 16:07:27: Pods: 17
2021-05-09 16:07:27: Running: 16
2021-05-09 16:07:27: Pending: 16
2021-05-09 16:07:27: Failed: 16
2021-05-09 16:07:27: Unknown: 16
2021-05-09 16:07:27: Completed: 16
2021-05-09 16:07:27: CrashLoopBackOff: 16
2021-05-09 16:07:27: NodeAffinity: 16
2021-05-09 16:07:27: ImagePullBackOff: 16
2021-05-09 16:07:27: NodeAffinity: 16
W0509 16:07:27.385906   13445 top_pod.go:140] Using json format to get metrics. Next release will switch to protocol-buffers, switch early by passing --use-protocol-buffers flag
W0509 16:07:27.494387   13450 top_pod.go:140] Using json format to get metrics. Next release will switch to protocol-buffers, switch early by passing --use-protocol-buffers flag
W0509 16:07:27.600589   13455 top_pod.go:140] Using json format to get metrics. Next release will switch to protocol-buffers, switch early by passing --use-protocol-buffers flag
2021-05-09 16:07:27: Collecting information about storage...
2021-05-09 16:07:28: Collecting information about cattle-dashboards
2021-05-09 16:07:30: Collecting information about cattle-logging-system
2021-05-09 16:07:32: Collecting information about cattle-monitoring-system
2021-05-09 16:07:35: Collecting information about cattle-system
2021-05-09 16:07:37: Collecting information about cis-operator-system
2021-05-09 16:07:39: Collecting information about fleet-system
2021-05-09 16:07:42: Collecting information about ingress-nginx
2021-05-09 16:07:44: Collecting information about kube-node-lease
2021-05-09 16:07:46: Collecting information about kube-public
2021-05-09 16:07:49: Collecting information about kube-system
2021-05-09 16:07:51: Collecting information about local-path-storage
2021-05-09 16:07:53: Collecting information about longhorn-system
2021-05-09 16:07:56: Deploying serviceaccount
2021-05-09 16:07:56: Deploying clusterrole
clusterrole.rbac.authorization.k8s.io/cluster-health-check created
2021-05-09 16:07:57: Deploying clusterrolebinding
2021-05-09 16:07:57: Deploying swiss-army-knife test containers
rancherlabs/swiss-army-knife:latest
daemonset.apps/swiss-army-knife created
2021-05-09 16:07:57: Waiting for rollout to complete...
Waiting for daemon set "swiss-army-knife" rollout to finish: 0 of 1 updated pods are available...
daemon set "swiss-army-knife" successfully rolled out
2021-05-09 16:08:22: Starting network overlay test
Error from server: error dialing backend: dial tcp: lookup ip-10-0-7-10 on 10.0.0.2:53: no such host
2021-05-09 16:08:23: Failure: pod swiss-army-knife-tlsc8 with the IP 10.42.0.12 on ip-10-0-7-10 cannot reach pod IP 10.42.0.12 on ip-10-0-7-10
2021-05-09 16:08:23: Finished network overlay test
2021-05-09 16:08:23: Starting cluster DNS test
Error from server: error dialing backend: dial tcp: lookup ip-10-0-7-10 on 10.0.0.2:53: no such host
2021-05-09 16:08:23: Failure: DNS is not working correctly for swiss-army-knife-tlsc8 on ip-10-0-7-10
2021-05-09 16:08:23: Finished DNS test
2021-05-09 16:08:23: Starting kubeapi test
Error from server: error dialing backend: dial tcp: lookup ip-10-0-7-10 on 10.0.0.2:53: no such host
2021-05-09 16:08:24: Failure: kubeapi is not working correctly for swiss-army-knife-tlsc8 on node ip-10-0-7-10 when connecting to kubeapi server 10.0.7.10
2021-05-09 16:08:24: Finished kubeapi test
2021-05-09 16:08:24: Cleaning up swiss-army-knife test containers
daemonset.apps "swiss-army-knife" deleted
2021-05-09 16:08:24: Deleting clusterrolebinding
clusterrolebinding.rbac.authorization.k8s.io "cluster-health-check" deleted
2021-05-09 16:08:24: Deleting clusterrole
clusterrole.rbac.authorization.k8s.io "cluster-health-check" deleted
2021-05-09 16:08:24: Deleting serviceaccount
serviceaccount "cluster-health-check" deleted
2021-05-09 16:08:24: Created /tmp/ip-10-0-7-10-2021-05-09_16_08_24.tar.gz
2021-05-09 16:08:24: Removing /tmp/tmp.LUZxYxiCy9
```

mattmattox commented:
@dnoland1 It works now in RKE2 clusters. It looks like the node selectors were failing and returning an error. So I added some error handling around that and cleaned up the output.

mattmattox reviewed (commented):
@leodotcloud Ready for review

mattmattox commented on cluster-health-check/cluster-health.sh:
```diff
@@ -0,0 +1,606 @@
+  numberOfPods=`cat $TMPDIR/pods/get-pods.json | jq -r '.items[] | .metadata.namespace + "/" + .metadata.name' | wc -l`
+
+  decho "Running"
+  cat $TMPDIR/pods/get-pods.json | jq -r '.items[] | select(.status.phase = "Running" ) | .metadata.namespace + "/" + .metadata.name' > $TMPDIR/pods/pods-Running 2>&1
```
Fixed

mattmattox commented:
@dnoland1 Ready for review

dnoland1 commented:
Tried on a RKE1 downstream cluster (v1.22.7) deployed with Rancher v2.6.4, but not working. Turned on debugging:

```
ip-10-0-98-10:~ # ./cluster-health.sh
+ SPACE=2048
+ PRIORITY_NICE=19
+ PRIORITY_IONICE=idle
+ TESTSONLY=0
+ CLEANUP=0
+ getopts :d:k:s:r:i:tcfhDyp opt
+ setup
++ mktemp -d
+ TMPDIR=/tmp/tmp.JvS1v44qVd
+ techo 'Created /tmp/tmp.JvS1v44qVd'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo '2022-04-11 17:37:13: Created /tmp/tmp.JvS1v44qVd'
2022-04-11 17:37:13: Created /tmp/tmp.JvS1v44qVd
+ '[' -z ']'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo -n '2022-04-11 17:37:13: Detecting available commands... '
2022-04-11 17:37:13: Detecting available commands... ++ command -v renice
+ renice -n 19 1023
+ echo -n 'renice '
renice ++ command -v ionice
+ ionice -c idle -p 1023
+ echo ionoice
ionoice
+ disk-space
++ df -m /tmp/tmp.JvS1v44qVd
++ awk '{ print $4 }'
++ tail -n 1
+ AVAILABLE=11096
+ '[' 11096 -lt 2048 ']'
+ '[' -n '' ']'
+ dependencies-check
+ hash jq
+ verify-access
+ techo 'Verifying cluster access'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo '2022-04-11 17:37:13: Verifying cluster access'
2022-04-11 17:37:13: Verifying cluster access
+ [[ ! -z '' ]]
+ [[ ! -z '' ]]
+ [[ ! -z '' ]]
++ command -v k3s
++ command -v docker
+ decho 'Detected that we'\''re running on a k8s nodes with a Rancher server pod'
+ [[ ! -z '' ]]
++ docker ps
++ head -1
++ grep k8s_rancher_rancher
++ cut '-d ' -f1
+ DOCKER_ID=
+ KUBECTL_CMD='docker exec  kubectl'
+ docker exec kubectl cluster-info
+ techo 'Can not access cluster'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo '2022-04-11 17:37:13: Can not access cluster'
2022-04-11 17:37:13: Can not access cluster
+ exit 1
```

Also failed on RKE2:
```
ip-10-0-98-12:~ # ./cluster-health.sh
+ SPACE=2048
+ PRIORITY_NICE=19
+ PRIORITY_IONICE=idle
+ TESTSONLY=0
+ CLEANUP=0
+ getopts :d:k:s:r:i:tcfhDyp opt
+ setup
++ mktemp -d
+ TMPDIR=/tmp/tmp.F1dWGyZacd
+ techo 'Created /tmp/tmp.F1dWGyZacd'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo '2022-04-11 17:40:28: Created /tmp/tmp.F1dWGyZacd'
2022-04-11 17:40:28: Created /tmp/tmp.F1dWGyZacd
+ '[' -z ']'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo -n '2022-04-11 17:40:28: Detecting available commands... '
2022-04-11 17:40:28: Detecting available commands... ++ command -v renice
+ renice -n 19 24768
+ echo -n 'renice '
renice ++ command -v ionice
+ ionice -c idle -p 24768
+ echo ionoice
ionoice
+ disk-space
++ df -m /tmp/tmp.F1dWGyZacd
++ awk '{ print $4 }'
++ tail -n 1
+ AVAILABLE=8080
+ '[' 8080 -lt 2048 ']'
+ '[' -n '' ']'
+ dependencies-check
+ hash jq
+ verify-access
+ techo 'Verifying cluster access'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo '2022-04-11 17:40:28: Verifying cluster access'
2022-04-11 17:40:28: Verifying cluster access
+ [[ ! -z '' ]]
+ [[ ! -z '' ]]
+ [[ ! -z '' ]]
++ command -v k3s
++ command -v docker
+ decho 'Detected that we'\''re running on a k8s nodes with a Rancher server pod'
+ [[ ! -z '' ]]
++ docker ps
++ head -1
++ cut '-d ' -f1
++ grep k8s_rancher_rancher
+ DOCKER_ID=
+ KUBECTL_CMD='docker exec  kubectl'
+ docker exec kubectl cluster-info
+ techo 'Can not access cluster'
++ timestamp
++ date '+%Y-%m-%d %H:%M:%S'
+ echo '2022-04-11 17:40:28: Can not access cluster'
2022-04-11 17:40:28: Can not access cluster
+ exit 1
```
Worked fine on k3s:
```
ip-10-0-98-11:~ # ./cluster-health.sh
2022-04-11 17:32:22: Created /tmp/tmp.ky97p5PZXQ
2022-04-11 17:32:22: Detecting available commands... renice ionoice
2022-04-11 17:32:22: Verifying cluster access
2022-04-11 17:32:23: Cluster access has been verified
2022-04-11 17:32:23: Collecting information about the cluster...
2022-04-11 17:32:25: Detecting the cluster provider...
2022-04-11 17:32:25: Collecting information about the nodes...
2022-04-11 17:32:28: Gathering node counts...
2022-04-11 17:32:30: Node Summary Report
2022-04-11 17:32:30: Total number of nodes: 1
2022-04-11 17:32:30: etcd: 1
2022-04-11 17:32:30: controlplane: 0
2022-04-11 17:32:30: worker: 1
2022-04-11 17:32:30: master: 1
2022-04-11 17:32:30: Pods Per Node
2022-04-11 17:32:31: ip-10-0-98-11 12
2022-04-11 17:32:31: Collecting information about the pods...
2022-04-11 17:32:34: Pod Summary Report
2022-04-11 17:32:34: Pods: 11
2022-04-11 17:32:34: Running: 7
2022-04-11 17:32:34: Pending: 0
2022-04-11 17:32:34: Failed: 0
2022-04-11 17:32:34: Unknown: 0
2022-04-11 17:32:34: Completed: 0
2022-04-11 17:32:34: CrashLoopBackOff: 0
2022-04-11 17:32:34: NodeAffinity: 0
2022-04-11 17:32:34: ImagePullBackOff: 0
2022-04-11 17:32:34: NodeAffinity: 0
2022-04-11 17:32:35: Collecting information about storage...
2022-04-11 17:32:37: Collecting information about cattle-dashboards
2022-04-11 17:32:46: Collecting information about cattle-logging-system
2022-04-11 17:32:54: Collecting information about cattle-monitoring-system
2022-04-11 17:33:03: Collecting information about cattle-system
2022-04-11 17:33:12: Collecting information about cis-operator-system
2022-04-11 17:33:21: Collecting information about fleet-system
2022-04-11 17:33:30: Collecting information about ingress-nginx
2022-04-11 17:33:38: Collecting information about kube-node-lease
2022-04-11 17:33:47: Collecting information about kube-public
2022-04-11 17:33:56: Collecting information about kube-system
2022-04-11 17:34:06: Collecting information about local-path-storage
2022-04-11 17:34:14: Collecting information about longhorn-system
2022-04-11 17:34:23: Deploying serviceaccount
2022-04-11 17:34:23: Deploying clusterrole
clusterrole.rbac.authorization.k8s.io/cluster-health-check created
2022-04-11 17:34:24: Deploying clusterrolebinding
2022-04-11 17:34:24: Deploying swiss-army-knife test containers
rancherlabs/swiss-army-knife:latest
daemonset.apps/swiss-army-knife created
2022-04-11 17:34:25: Waiting for rollout to complete...
Waiting for daemon set "swiss-army-knife" rollout to finish: 0 of 1 updated pods are available...
daemon set "swiss-army-knife" successfully rolled out
2022-04-11 17:34:44: Starting network overlay test
2022-04-11 17:34:47: Success: pod swiss-army-knife-r5949 with the IP 10.42.0.15 on ip-10-0-98-11 can reach pod IP 10.42.0.15 on ip-10-0-98-11
2022-04-11 17:34:47: Finished network overlay test
2022-04-11 17:34:47: Starting cluster DNS test
2022-04-11 17:34:48: Success: DNS is working correctly for swiss-army-knife-r5949 on ip-10-0-98-11
2022-04-11 17:34:48: Finished DNS test
2022-04-11 17:34:48: Starting kubeapi test
2022-04-11 17:34:50: Success: kubeapi is working correctly for swiss-army-knife-r5949 on node ip-10-0-98-11 when connecting to kubeapi server 10.0.98.11
2022-04-11 17:34:50: Finished kubeapi test
2022-04-11 17:34:50: Cleaning up swiss-army-knife test containers
daemonset.apps "swiss-army-knife" deleted
2022-04-11 17:34:50: Deleting clusterrolebinding
clusterrolebinding.rbac.authorization.k8s.io "cluster-health-check" deleted
2022-04-11 17:34:50: Deleting clusterrole
clusterrole.rbac.authorization.k8s.io "cluster-health-check" deleted
2022-04-11 17:34:51: Deleting serviceaccount
serviceaccount "cluster-health-check" deleted
2022-04-11 17:34:51: Created /tmp/ip-10-0-98-11-2022-04-11_17_34_51.tar.gz
2022-04-11 17:34:51: Removing /tmp/tmp.ky97p5PZXQ
```
